services/venice-controller/src/main/java/com/linkedin/venice/controller/VeniceHelixAdmin.java

```diff
@@ -3320,6 +3308,49 @@ public void setStorePartitionCount(String clusterName, String storeName, int par
     });
   }
 
+  void preCheckStorePartitionCountUpdate(String clusterName, Store store, int newPartitionCount) {
+    String errorMessagePrefix = "Store update error for " + store.getName() + " in cluster: " + clusterName + ": ";
+    VeniceControllerClusterConfig clusterConfig = getHelixVeniceClusterResources(clusterName).getConfig();
+    if (store.isHybrid() && store.getPartitionCount() != newPartitionCount) {
+      // Allow the update if partition count is not configured and the new partition count matches RT partition count
+      if (store.getPartitionCount() == 0) {
+        TopicManager topicManager;
+        if (isParent()) {
+          // RT might not exist in parent colo. Get RT partition count from a child colo.
+          String childDatacenter = clusterConfig.getChildDatacenters().split(",\\s*")[0];
+          topicManager = getTopicManager(
+              Pair.create(
+                  multiClusterConfigs.getChildDataCenterKafkaUrlMap().get(childDatacenter),
+                  multiClusterConfigs.getChildDataCenterKafkaZkMap().get(childDatacenter)));
+        } else {
+          topicManager = getTopicManager();
+        }
+        String realTimeTopic = Version.composeRealTimeTopic(store.getName());
+        if (topicManager.containsTopic(realTimeTopic)
+            && topicManager.partitionsFor(realTimeTopic).size() == newPartitionCount) {
+          LOGGER.info("Allow updating store " + store.getName() + " partition count to " + newPartitionCount);
+          return;
+        }
+      }
+      String errorMessage = errorMessagePrefix + "Cannot change partition count for this hybrid store";
+      LOGGER.error(errorMessage);
+      throw new VeniceHttpException(HttpStatus.SC_BAD_REQUEST, errorMessage, ErrorType.INVALID_CONFIG);
+    }
+
+    int maxPartitionNum = clusterConfig.getMaxNumberOfPartition();
+    if (newPartitionCount > maxPartitionNum) {
+      String errorMessage =
+          errorMessagePrefix + "Partition count: " + newPartitionCount + " should be less than max: " + maxPartitionNum;
+      LOGGER.error(errorMessage);
+      throw new VeniceHttpException(HttpStatus.SC_BAD_REQUEST, errorMessage, ErrorType.INVALID_CONFIG);
+    }
+    if (newPartitionCount < 0) {
+      String errorMessage = errorMessagePrefix + "Partition count: " + newPartitionCount + " should NOT be negative";
+      LOGGER.error(errorMessage);
+      throw new VeniceHttpException(HttpStatus.SC_BAD_REQUEST, errorMessage, ErrorType.INVALID_CONFIG);
+    }
```

Comment:
Shall we move the sanity checks here to the beginning part of this function, otherwise it could return early @3332 without passing these checks?

Reply:
The partition count of RT could not be negative. For existing hybrid stores that we want to backfill partition count, their RT partition counts do not exceed the max either. So if the provided partition count equals to RT partition count, it should pass the other two checks.

Supposing that max partition count is initially small, and then is increased and decreased again, an existing hybrid store could have RT that has more partition counts than the max. However, since the store is already in use, we should allow updating the partition count to match RT partition count. Otherwise, keeping the partition count as 0 will have the risk of RT and VT partition counts mismatch.

Reply:
That is an interesting case regarding the maxPartitionNum condition check. Thanks for sharing it and please add some comments in the code to explain the intuition why we allow it to pass, in order to handle this corner case.